### PR TITLE
[3146] Handle course starting in the future for deferring

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/WjPPrUDt/3146-m-course-starts-in-the-future-defer
- https://trello.com/c/qxWQ8yi5/3164-s-reinstating-a-deferred-trainee

### Changes proposed in this pull request

- Introduce a new defer flow if the course starts in the future
- Trainee can be deferred in the scenario above but date options aren't shown to provider
- If reinstated, the trainee's `commencement_date` is marked the same as the `reinstate_date`

### Guidance to review

This flow hinges on the trainee's commencement date not being set as the course start date is in the future. There are tickets removing capturing this from the draft sections and after the trn has been requested. In that scenario, the commencement date is only requested if the course start date is on the current day or in the past. Therefore this field will show as missing data (spoken with Ed) when the course start date is due.

Test with any of these trainees:

- Change the course start date to be in the future
- The record details component should show a `Course has not started` message
- Defer the trainee, the trainee should be deferred with no date set
- Reinstate the trainee, (you'll have to set the reinstated date on or after the course start date due to validations), the trainee's commencement date should match the reinstated date

### CODE REVIEW

**Comment by darokel** [comment]
[[M] Course starts in the future — Defer](https://trello.com/c/WjPPrUDt/3146-m-course-starts-in-the-future-defer)

**Comment by darokel** [comment]
[[S] Reinstating a deferred trainee](https://trello.com/c/qxWQ8yi5/3164-s-reinstating-a-deferred-trainee)

File: app/components/record_details/view.rb [hunk 1]
```diff
@@ -93,7 +93,9 @@ def record_created_row
     end
 
     def trainee_start_date_row
-      mappable_field(trainee_start_date, t(".trainee_start_date"), edit_trainee_start_date_path(trainee))
+      change_link = trainee.starts_course_in_the_future? ? nil : edit_trainee_start_date_path(trainee)
+
+      mappable_field(trainee_start_date, t(".trainee_start_date"), change_link)
```
Comment (AbigailMcP) [inline app/components/record_details/view.rb:98]: Might need to be agreed with @avinhurry and this PR - https://github.com/DFE-Digital/register-trainee-teachers/pull/1724/files

I _think_ you're both achieving the same thing!

File: app/components/record_details/view.rb [hunk 2]
```diff
@@ -103,7 +105,10 @@ def render_text_with_hint(date)
     end
 
     def trainee_start_date
-      trainee.commencement_date.present? ? date_for_summary_view(trainee.commencement_date) : nil
+      return t(".course_not_started") if trainee.starts_course_in_the_future?
```
Comment (AbigailMcP) [inline app/components/record_details/view.rb:108]: Oh no.. and i think @mooktakim has a ticket for displaying the trainee start date! 😓

File: config/locales/en.yml [hunk 3]
```diff
@@ -59,6 +59,7 @@ en:
       last_updated: Last updated
       record_created: Record created
       trainee_start_date: Trainee start date
+      course_not_started: Course has not started
```
Comment (AbigailMcP) [inline config/locales/en.yml:62]: I think this might now be "ITT has not started" cc @ralph-hawkins

**Review by AbigailMcP — COMMENTED** [review]
This might not be a problem.. but if I've withdrawn a trainee who's course start date is in the future, then I can't reinstate them until the course starts cc @ralph-hawkins 

Screenshot 2021-11-11 at 12 40 13

**Comment by darokel** [comment]
> This might not be a problem.. but if I've withdrawn a trainee who's course start date is in the future, then I can't reinstate them until the course starts cc @ralph-hawkins
> 
> Screenshot 2021-11-11 at 12 40 13

Yeah there are validations preventing that otherwise dttp will blow up if these dates are set before the course start date. Seems like you have to either reinstate to the same start date or after the course starts

File: app/components/record_details/view.rb (hunk 1 above)
Comment (darokel) [reply app/components/record_details/view.rb:98]: filthy overlap-ses

File: config/locales/en.yml (hunk 3 above)
Comment (darokel) [reply config/locales/en.yml:62]: i was holding off on that until it was actually implemented, we're still showing course start date at the moment?

**Review by AbigailMcP — APPROVED** [review]

**Comment by AbigailMcP** [comment]
> Yeah there are validations preventing that otherwise dttp will blow up if these dates are set before the course start date. Seems like you have to either reinstate to the same start date or after the course starts

Got it! Just wanted to flag as it felt a bit jarring when i was testing 👍

File: app/components/record_details/view.rb (hunk 2 above)
Comment (mooktakim) [reply app/components/record_details/view.rb:108]: Yep, this method is already different on main based on what @archferns merged recently. 
Here's my card:
https://trello.com/c/3DnYjTL4/3161-trainee-start-date-row-record-view

File: config/locales/en.yml (hunk 3 above)
Comment (ralph-hawkins) [reply config/locales/en.yml:62]: I think right now prod shows either "course" or "ITT" depending on the type of route. Its in the backlog to only use "ITT", but until then we should use either "course" or "ITT" consistently for that route.

**Comment by ralph-hawkins** [comment]
> This might not be a problem.. but if I've withdrawn a trainee who's course start date is in the future, then I can't reinstate them until the course starts cc @ralph-hawkins

We need to allow the trainee to be reinstated. If the course hasn't started then can we:
- skip the "When did the trainee return?" question
- record the trainee as still not having started in Register
- store the start date in DTTP as the same as the course start date (which we're doing anyway)

This way to DTTP it will look like the same as all the other trainees who never withdrew.

**Pull request merged**